### ISSUE TITLE
open-iscsi-2.0.877 fails to build on Void Linux: requires systemd

### ISSUE THREAD

**Issue by maxice8** [open-iscsi/open-iscsi#151]
Fails with

```
/usr/bin/ld: libopeniscsiusr.so.0.2.0: undefined reference to `kmod_module_probe_insert_module'
/usr/bin/ld: libopeniscsiusr.so.0.2.0: undefined reference to `kmod_module_unref'
/usr/bin/ld: libopeniscsiusr.so.0.2.0: undefined reference to `kmod_new'
/usr/bin/ld: libopeniscsiusr.so.0.2.0: undefined reference to `kmod_unref'
/usr/bin/ld: libopeniscsiusr.so.0.2.0: undefined reference to `kmod_load_resources'
/usr/bin/ld: libopeniscsiusr.so.0.2.0: undefined reference to `kmod_module_new_from_name'
collect2: error: ld returned 1 exit status
```

After adding an [Arch Linux patch](https://git.archlinux.org/svntogit/community.git/tree/trunk/0001-libkmod.patch?h=packages/open-iscsi) to add `-lkmod` it now fails with

```
iscsid.c:37:10: fatal error: systemd/sd-daemon.h: No such file or directory
 #include <systemd/sd-daemon.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

**Comment by gonzoleeman** [comment]
The arch Linux patch you point at has been fixed in open-iscsi already, a bit differently. See commits 623a81123c494 and 623a81123c494. Evidently the sources from which you are building are not current.

As far as the include of sd-daemon.h failing, where is your sd-daemon.h located? It's in /usr/include/systemd on my system and others I've checked, and it's part of a systemd-devel package -- the same one that supplies libsystemd.so.

**Comment by maxice8** [comment]
> The arch Linux patch you point at has been fixed in open-iscsi already, a bit differently. See commits 623a811 and 623a811. Evidently the sources from which you are building are not current.

last tagged release 2.0.877

> As far as the include of sd-daemon.h failing, where is your sd-daemon.h located? It's in /usr/include/systemd on my system and others I've checked, and it's part of a systemd-devel package -- the same one that supplies libsystemd.so.

Void doesn't use systemd as it doesn't support musl libc and introduces dependency cycles with dbus and polkit making bootstrap impossible without various loops of building.

**Comment by gonzoleeman** [comment]
I see. So there are linux distros that eschew systemd. That presents a problem. The open-iscsi project moved from sysV-init support, to dual support, to systemd-only support over the last few years.

I will consider a pull request for changes to make the systemd stuff ifdef-able, as it was for a while, as long as the default is to use systemd.

I think the _right_ way to deal with this might be to move to autoconfig, which could (among other things) figure out if systemd is present or not and configure ifdefs appropriately.

**Comment by maxice8** [comment]
> I will consider a pull request for changes to make the systemd stuff ifdef-able, as it was for a while, as long as the default is to use systemd.

Sounds good, i will try to arrange some time to make proper patches.

> I think the right way to deal with this might be to move to autoconfig, which could (among other things) figure out if systemd is present or not and configure ifdefs appropriately.

If switching to a new build system is tempting i reccomend [meson](https://mesonbuild.com/)

**Comment by gonzoleeman** [comment]
Renamed issue to better describe the problem.

Part of the problem with doing away with systemd now is that the pre-systemd init file was a nightmare, and was different for each distribution. So if you re-create that, you will also have to help support it. It might be easier for you to have a distro-specific patch, but I'll consider making systemd usage optional, as long as it's not *too* ugly.

**Comment by gonzoleeman** [comment]
This might be a duplicate of issue #149

**Comment by maxice8** [comment]
> Part of the problem with doing away with systemd now is that the pre-systemd init file was a nightmare, 

Understandable, i don't want to do away with systemd, i don't really care about it, i just don't want it be able to be switched off for distros like Void Linux and Alpine Linux that don't use systemd.

> So if you re-create that, you will also have to help support it.

I have no plans to make upstream stop using systemd, my objective is to work on some patches (when i get time)  so that systemd is on by default but optional.

> It might be easier for you to have a distro-specific patch

We have some patches to make systemd optional on but Void Linux as a distro, like arch, is **very** averse from deviating from upstream.

**Comment by gonzoleeman** [comment]
Ok, I got tired of waiting for a pull request on this one, and since I was working on systemd for iscsiuio I added a pull request to fix this. See pull request #164

**Comment by gonzoleeman** [comment]
Should have been resolved with issue #164
